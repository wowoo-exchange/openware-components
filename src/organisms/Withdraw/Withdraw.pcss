.cr-withdraw {
  padding: 10px;

  &__divider {
    background-color: var(--withdraw-divider-border-color);
    width: 1px;
    margin-left: calc(var(--gap) * 3.45);
    min-height: calc(var(--gap) * 3.2);

    &-one {
      height: calc(var(--gap) * 6.4);
    }

    &-two {
      position: absolute;
      height: calc(var(--gap) * 13);
    }
  }

  &__summary-field {
    margin-left: calc(var(--gap) * 2.9);
    margin-top: calc(var(--gap) * 4.5);

    .cr-summary-field-message {
      color: var(--withdraw-summary-field-message-color);
    }

    .cr-summary-field-amount {
      color: var(--withdraw-summary-field-amount-color);
    }
  }

  &__deep {
    text-align: right;
  }

  &__button {
    background: var(--withdraw-button-bg);
    color: #fff;
    font-size: calc(var(--gap) * 3);
    font-style: normal;
    font-weight: bold;
    height: calc(var(--gap) * 6.5);
    line-height: normal;
    width: calc(var(--gap) * 30.8);
  }
}
