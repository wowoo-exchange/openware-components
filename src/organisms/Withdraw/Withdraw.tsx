import classnames from 'classnames';
import * as React from 'react';
import { Button, Input, SummaryField } from '../../atoms';
import { Decimal } from '../../format';

interface WithdrawProps {
    /**
     * Value for setting correct symbol for summary field component. Could be 'rectangle', 'circle' or 'empty-circle'
     */
    borderItem?: string;
    /**
     * Value for setting current currency for withdraw
     */
    currency: string;
    /**
     * Fee for defined currency
     * @default 0
     */
    fee: number;
    /**
     * Function for withdraw event
     */
    onClick: (amount: number, rid: string, otpCode: string) => void;
    /**
     * The precision for numbers
     */
    fixed: number;
    /**
     * Additional class name. By default element receives `cr-withdraw` class
     * @default empty
     */
    className?: string;
    /**
     * Renders Google Authenticator code input
     */
    twoFactorAuthRequired?: boolean;
}

interface WithdrawState {
    /**
     * Value for current address for withdraw
     * @default ''
     */
    address: string;
    /**
     * Value for current amount for withdraw
     * @default 0
     */
    amount: number | string;
    /**
     * Google Authenticator 6-digit code
     */
    otpCode: string;
    /**
     * Total value for withdraw
     */
    total: number;
}

/**
 * Component with for for withdraw.
 */
class Withdraw extends React.Component<WithdrawProps, WithdrawState> {
    public state = {
        address: '',
        amount: '',
        otpCode: '',
        total: 0,
    };

    public render() {
        const {
            address,
            amount,
            total,
        } = this.state;
        const {
            borderItem,
            className,
            currency,
            twoFactorAuthRequired,
        } = this.props;
        const cx = classnames('cr-withdraw', className);
        const lastDividerClassName = classnames('cr-withdraw__divider', {
            'cr-withdraw__divider-one': twoFactorAuthRequired,
            'cr-withdraw__divider-two': !twoFactorAuthRequired,
        });

        const formattedCurrency = currency.toUpperCase();
        return (
            <div className={cx}>
                <div className="cr-withdraw-column">
                    <form>
                        <fieldset className="cr-withdraw__input">
                            <legend>
                                {formattedCurrency} "Withdrawal Address"
                            </legend>
                            <Input
                                className="cr-input-block__input"
                                type="text"
                                placeholder="Address"
                                value={address}
                                onChangeValue={this.handleChangeInputAddress}
                            />
                        </fieldset>
                    </form>
                    <div className="cr-withdraw__divider cr-withdraw__divider-one" />
                    <form>
                        <fieldset className="cr-withdraw__input">
                            <legend>
                                Withdrawal Amount
                            </legend>
                            <Input
                                className="cr-input-block__input"
                                type="number"
                                placeholder="0"
                                value={amount}
                                onChangeValue={this.handleChangeInputAmount}
                            />
                        </fieldset>
                    </form>
                    <div className={lastDividerClassName} />
                    {twoFactorAuthRequired && this.renderOtpCodeInput()}
                </div>
                <div className="cr-withdraw-column">
                    <div>
                        <SummaryField
                            className="cr-withdraw__summary-field "
                            message="Fee"
                            content={this.renderFee()}
                            borderItem={borderItem}
                        />
                        <SummaryField
                            className="cr-withdraw__summary-field"
                            message="Total Withdraw Amount"
                            content={this.renderTotal()}
                            borderItem={borderItem}
                        />
                    </div>
                    <div className="cr-withdraw__deep">
                        <Button
                            className="cr-withdraw__button"
                            label="WITHDRAW"
                            onClick={this.handleClick}
                            disabled={Number(total) <= 0 || !Boolean(address)}
                        />
                    </div>
                </div>
            </div>
        );
    }

    private renderFee = () => {
        const { fee, fixed, currency } = this.props;
        return (
            <span>
                <Decimal fixed={fixed}>{fee.toString()}</Decimal> {currency.toUpperCase()}
            </span>
            );
    }

    private renderTotal = () => {
        const total = this.state.total;
        const { fixed, currency } = this.props;
        return total ? (
            <span>
                <Decimal fixed={fixed}>{total.toString()}</Decimal> {currency.toUpperCase()}
            </span>) :
            <span>0 {currency.toUpperCase()}</span>;
    }

    private renderOtpCodeInput = () => {
        const { otpCode } = this.state;
        return (
            <React.Fragment>
                <form>
                    <fieldset className="cr-withdraw__input">
                        <legend>
                            6-digit GAuthenticator Code
                        </legend>
                        <Input
                            type="text"
                            className="cr-input-block__input"
                            placeholder="XXXXXX"
                            value={otpCode}
                            onChangeValue={this.handleChangeInputOtpCode}
                        />
                    </fieldset>
                </form>
                <div className="cr-withdraw__divider cr-withdraw__divider-two" />
            </React.Fragment>
        );
    }

    private handleClick = () => this.props.onClick(
        parseInt(this.state.amount, 10),
        this.state.address,
        this.state.otpCode,
    );

    private handleChangeInputAmount = (text: string) => {
        const value: number = parseFloat(text);
        const total: number = value - this.props.fee;
        if (total < 0) {
            this.setTotal(0);
        } else {
            this.setTotal(total);
        }
        this.setState({
            amount: value,
        });
    };

    private setTotal = (value: number) => {
        this.setState({
            total: value,
        });
    };

    private handleChangeInputAddress = (text: string) => {
        this.setState({
            address: text,
        });
    };

    private handleChangeInputOtpCode = (otpCode: string) => {
        this.setState({ otpCode });
    }
}

export {
    Withdraw,
    WithdrawProps,
    WithdrawState,
};
