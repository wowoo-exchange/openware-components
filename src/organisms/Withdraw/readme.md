Withdraw example:

```js
const handleClick = () => { alert('Withdraw'); }

<div className="bg-dark">
  <Withdraw
    currency="BTC"
    fee={0.000111}
    onClick={handleClick}
    fixed={8}
    borderItem="empty-circle"
  />
</div>
```
