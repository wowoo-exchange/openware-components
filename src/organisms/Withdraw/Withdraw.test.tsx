import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { SinonSpy, spy } from 'sinon';
import { Withdraw, WithdrawProps } from './Withdraw';

const defaultProps: WithdrawProps = {
    currency: 'BTC',
    fee: 0.1,
    onClick: spy(),
    fixed: 8,
};

const setup = (props: Partial<WithdrawProps> = {}) =>
    shallow(<Withdraw {...{ ...defaultProps, ...props }} />);

describe('Withdraw', () => {
    it('should render', () => {
        const wrapper = setup();
        expect(wrapper).toMatchSnapshot();
    });

    it('should have correct className', () => {
        const wrapper = setup();
        expect(wrapper.hasClass('cr-withdraw')).toBeTruthy();
    });

    it('should pass along supplied className', () => {
        const className = 'some-class';
        const wrapper = setup({ className });
        expect(wrapper.hasClass(className)).toBeTruthy();
    });

    it('should change address in state', () => {
        const wrapper = setup();
        const children = wrapper.children();
        const address = children.first().children().first().children().first().children().last();
        const props = address.props();
        props.onChangeValue('potato');
        expect(wrapper.state()).toEqual({address: 'potato', amount: '', total: 0, otpCode: ''});
    });

    it('should change amount and total in state', () => {
        const wrapper = setup();
        const children = wrapper.children();
        // tslint:disable-next-line:no-magic-numbers
        const address = children.first().children().at(2).children().first().children().last();
        const props = address.props();
        props.onChangeValue(1);
        expect(wrapper.state()).toEqual({address: '', amount: 1, total: 0.9, otpCode: ''});
    });

    it('should change amount and total in state', () => {
        const wrapper = setup();
        const children = wrapper.children();
        // tslint:disable-next-line:no-magic-numbers
        const address = children.first().children().at(2).children().first().children().last();
        const props = address.props();
        // tslint:disable-next-line:no-magic-numbers
        props.onChangeValue(-5);
        expect(wrapper.state()).toEqual({address: '', amount: -5, total: 0, otpCode: ''});
    });

    it('shouldn\'t handle submit function before ', () => {
        // tslint:disable: no-shadowed-variable
        const wrapper = mount(<Withdraw {...{ ...defaultProps}} />);
        wrapper.find('input').last().simulate('click');
        expect((defaultProps.onClick as SinonSpy).calledOnceWith()).toBeFalsy();
        wrapper.unmount();
    });

    it('should handle submit function after filling data', () => {
        // tslint:disable: no-shadowed-variable
        const wrapper = mount(<Withdraw {...{ ...defaultProps}} />);
        wrapper.find('input').first().simulate('change', {target: { value: '123' }});
        wrapper.find('input').at(1).simulate('change', {target: { value: '123' }});
        wrapper.find('input').last().simulate('click');
        expect((defaultProps.onClick as SinonSpy).calledOnceWith()).toBeTruthy();
        wrapper.unmount();
    });
});
