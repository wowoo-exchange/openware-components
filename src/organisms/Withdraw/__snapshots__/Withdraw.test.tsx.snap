// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Withdraw should render 1`] = `
<div
  className="cr-withdraw"
>
  <div
    className="cr-withdraw-column"
  >
    <form>
      <fieldset
        className="cr-withdraw__input"
      >
        <legend>
          BTC
           "Withdrawal Address"
        </legend>
        <Input
          className="cr-input-block__input"
          onChangeValue={[Function]}
          placeholder="Address"
          type="text"
          value=""
        />
      </fieldset>
    </form>
    <div
      className="cr-withdraw__divider cr-withdraw__divider-one"
    />
    <form>
      <fieldset
        className="cr-withdraw__input"
      >
        <legend>
          Withdrawal Amount
        </legend>
        <Input
          className="cr-input-block__input"
          onChangeValue={[Function]}
          placeholder="0"
          type="number"
          value=""
        />
      </fieldset>
    </form>
    <div
      className="cr-withdraw__divider cr-withdraw__divider-two"
    />
  </div>
  <div
    className="cr-withdraw-column"
  >
    <div>
      <SummaryField
        className="cr-withdraw__summary-field "
        content={
          <span>
            <Decimal
              fixed={8}
            >
              0.1
            </Decimal>
             
            BTC
          </span>
        }
        message="Fee"
      />
      <SummaryField
        className="cr-withdraw__summary-field"
        content={
          <span>
            0 
            BTC
          </span>
        }
        message="Total Withdraw Amount"
      />
    </div>
    <div
      className="cr-withdraw__deep"
    >
      <Button
        className="cr-withdraw__button"
        disabled={true}
        label="WITHDRAW"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;
